packages/library/src/compute-limit-internal.ts

```diff
@@ -0,0 +1,187 @@
+     *         - either a recent blockhash lifetime or a nonce.
+     */
+    const isDurableNonceTransactionMessage = isDurableNonceTransaction(transactionMessage);
+    let compilableTransactionMessage;
+    if (isDurableNonceTransactionMessage || isTransactionMessageWithBlockhashLifetime(transactionMessage)) {
+        compilableTransactionMessage = transactionMessage;
+    } else {
+        compilableTransactionMessage = setTransactionMessageLifetimeUsingBlockhash(
+            INVALID_BUT_SUFFICIENT_FOR_COMPILATION_BLOCKHASH,
+            transactionMessage,
+        );
+    }
+    /**
+     * STEP 2: Ensure that the message has a `SetComputeLimit` instruction. The set compute limit
+     *         instruction itself consumes compute units, so it must be included in the simulation.
+     */
+    const existingSetComputeUnitLimitInstructionIndex =
+        transactionMessage.instructions.findIndex(isSetComputeLimitInstruction);
+    const maxComputeUnitLimitInstruction = createComputeUnitLimitInstruction(4294967295 /* U32::MAX */);
+    if (existingSetComputeUnitLimitInstructionIndex === -1) {
+        compilableTransactionMessage = appendTransactionMessageInstruction(
+            maxComputeUnitLimitInstruction,
+            compilableTransactionMessage,
+        );
+    } else {
+        const nextInstructions = [...compilableTransactionMessage.instructions];
+        nextInstructions.splice(existingSetComputeUnitLimitInstructionIndex, 1, maxComputeUnitLimitInstruction);
+        compilableTransactionMessage = Object.freeze({
+            ...compilableTransactionMessage,
+            instructions: nextInstructions,
+        } as typeof compilableTransactionMessage);
+    }
+    /**
+     * STEP 3: Simulate the transaction to measure its compute unit consumption.
+     */
+    const compiledTransaction = compileTransaction(compilableTransactionMessage);
+    const wireTransactionBytes = getBase64EncodedWireTransaction(compiledTransaction);
+    try {
+        const {
+            value: { unitsConsumed },
+        } = await rpc
+            .simulateTransaction(wireTransactionBytes, {
+                ...simulateConfig,
+                encoding: 'base64',
+                replaceRecentBlockhash: !isDurableNonceTransactionMessage,
+                sigVerify: false,
+            })
+            .send({ abortSignal });
+        if (unitsConsumed == null) {
+            // This should never be hit, because all RPCs should support `unitsConsumed` by now.
+            throw new SolanaError(SOLANA_ERROR__TRANSACTION__FAILED_TO_ESTIMATE_COMPUTE_LIMIT);
+        }
+        // FIXME: The simulation response returns compute units as a u64, but the `SetComputeLimit`
+        // instruction only accepts a u32. Until this changes, downcast it.
+        const downcastUnitsConsumed = unitsConsumed > 4294967295n ? 4294967295 : Number(unitsConsumed);
+        return downcastUnitsConsumed;
```

Comment:
I'm not sure how much knowledge we want to bake in here, but in practice if this is >1.4M it's an error

Reply:
I think I just want to let the server tell me, and if it tells me a wrong thing then oh well.